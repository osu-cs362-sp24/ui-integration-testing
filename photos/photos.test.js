/**
 * @jest-environment jsdom
 */

require("@testing-library/jest-dom")
const domTesting = require("@testing-library/dom")
const userEvent = require("@testing-library/user-event").default

const fs = require("fs")

function initDomFromFiles(htmlPath, jsPath) {
  const html = fs.readFileSync(htmlPath, "utf8")
  document.open()
  document.write(html)
  document.close()
  jest.isolateModules(function () {
    require(jsPath)
  })
}

test("photo card is created when URL and caption are provided", async function () {
  initDomFromFiles(
    __dirname + "/photos.html",
    __dirname + "/photos.js"
  )

  const urlInput = domTesting.getByLabelText(document, "Photo URL")
  const captionInput = domTesting.getByLabelText(document, "Caption")
  const addPhotoButton = domTesting.getByRole(document, "button")
  const photoCardList = domTesting.getByRole(document, "list")

  const user = userEvent.setup()
  await user.type(urlInput, "http://placekitten.com/480/480")
  await user.type(captionInput, "Kitty photo")
  await user.click(addPhotoButton)

  expect(photoCardList).not.toBeEmptyDOMElement()

  const photoCards = domTesting.queryAllByRole(document, "listitem")
  expect(photoCards).toHaveLength(1)

  const photoCard = photoCards[0]
  const img = domTesting.queryByRole(photoCard, "img")
  expect(img).toHaveAttribute("src", "http://placekitten.com/480/480")

  const caption = domTesting.queryByText(photoCard, "Kitty photo")
  expect(caption).not.toBeNull()
})

test("clears form when correctly submitted", async function () {
  initDomFromFiles(
    __dirname + "/photos.html",
    __dirname + "/photos.js"
  )

  const urlInput = domTesting.getByLabelText(document, "Photo URL")
  const captionInput = domTesting.getByLabelText(document, "Caption")
  const addPhotoButton = domTesting.getByRole(document, "button")

  const user = userEvent.setup()
  await user.type(urlInput, "http://placekitten.com/480/480")
  await user.type(captionInput, "Kitty photo")
  await user.click(addPhotoButton)

  expect(urlInput).not.toHaveValue()
  expect(captionInput).not.toHaveValue()
})

test("no photo card is added without a caption", async function () {
  initDomFromFiles(
    __dirname + "/photos.html",
    __dirname + "/photos.js"
  )

  const urlInput = domTesting.getByLabelText(document, "Photo URL")
  const addPhotoButton = domTesting.getByRole(document, "button")
  const photoCardList = domTesting.getByRole(document, "list")

  const user = userEvent.setup()
  await user.type(urlInput, "http://placekitten.com/480/480")
  await user.click(addPhotoButton)

  expect(photoCardList).toBeEmptyDOMElement()
  expect(urlInput).toHaveValue("http://placekitten.com/480/480")
})

test("photo card is created when URL and caption are provided", async function () {
  initDomFromFiles(
    __dirname + "/photos.html",
    __dirname + "/photos.js"
  )

  const urlInput = domTesting.getByLabelText(document, "Photo URL")
  const captionInput = domTesting.getByLabelText(document, "Caption")
  const addPhotoButton = domTesting.getByRole(document, "button")
  const photoCardList = domTesting.getByRole(document, "list")

  const user = userEvent.setup()
  await user.type(urlInput, "http://placekitten.com/480/480")
  await user.type(captionInput, "Kitty photo")
  await user.click(addPhotoButton)

  await user.type(urlInput, "http://placekitten.com/320/320")
  await user.type(captionInput, "Kitty photo #3")
  await user.click(addPhotoButton)

  expect(photoCardList).toMatchSnapshot()
})
