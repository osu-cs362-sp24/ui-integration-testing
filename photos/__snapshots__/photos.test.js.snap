// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`photo card is created when URL and caption are provided 1`] = `
<ul
  id="photo-card-list"
>
  <li
    class="photo-card"
  >
    <img
      alt="Kitty photo"
      src="http://placekitten.com/480/480"
    />
    <p>
      Kitty photo
    </p>
  </li>
  <li
    class="photo-card"
  >
    <img
      alt="Kitty photo #3"
      src="http://placekitten.com/320/320"
    />
    <p>
      Kitty photo #3
    </p>
  </li>
</ul>
`;
